### PR TITLE
throw when different client frameworks are used for microfrontends

### LINKED ISSUE #18200: JDL validation should catch different clientFrameworks when using microfrontends
##### **Overview of the issue**

If you create a JDL that has different client frameworks in microservices, you won't know this fails until runtime. For example, if you modified [reactive-mf.jdl](https://github.com/jhipster/jdl-samples/blob/main/reactive-mf.jdl) to have something like the following:

```
application {
  config {
    ...
    clientFramework angular
    ...
  }
  ...
}

application {
  config {
    ...
    clientFramework react
    ...
  }
  ...
}

application {
  config {
    ...
    clientFramework vue
    ...
  }
  ...
}

...
```

If you generate apps with the above JDL, the gateway will have an error in the browser console and menu items aren't shown.

```
navbar.component.ts:73 Error loading store entities ScriptExternalLoadError: Loading script failed.
(error: http://localhost:8080/services/store/remoteEntry.js)
    at Object.908 (:8080/main.js:46:25)
    at __webpack_require__ (:8080/main.js:83:41)
    at initExternal (:8080/main.js:382:28)
    at Function.__webpack_require__.I (:8080/main.js:411:15)
    at :8080/main.js:543:47
    at Object.3037 (:8080/main.js:597:26)
    at :8080/main.js:670:56
    at Array.forEach (<anonymous>)
    at Object.__webpack_require__.f.consumes (:8080/main.js:653:36)
    at :8080/main.js:157:40
navbar.component.ts:62 Error loading blog entities Error: Module "./entity-navbar-items" does not exist in container.
while loading "./entity-navbar-items" from 1527
    at :8080/services/blog/remoteEntry.js:26:11
    at ZoneDelegate.invoke (:8080/polyfills.js:434:30)
    at Object.onInvoke (:8080/node_modules_angular_core_fesm2015_core_mjs.js:25904:33)
    at ZoneDelegate.invoke (:8080/polyfills.js:433:56)
    at Zone.run (:8080/polyfills.js:194:47)
    at :8080/polyfills.js:1352:38
    at ZoneDelegate.invokeTask (:8080/polyfills.js:468:35)
    at Object.onInvokeTask (:8080/node_modules_angular_core_fesm2015_core_mjs.js:25891:33)
    at ZoneDelegate.invokeTask (:8080/polyfills.js:467:64)
    at Zone.runTask (:8080/polyfills.js:239:51)
```

##### **Motivation for or Use Case**

We should catch issues we know about at generation time rather than runtime.

### PR DESCRIPTION
**Pull request by emilpaw**
Fix #18200

---

Please make sure the below checklist is followed for Pull Requests.

- [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [ ] Tests are added where necessary
- [x] The JDL part is updated if necessary
- [x] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [x] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

File: cli/import-jdl.js [hunk 1]
```diff
@@ -440,6 +440,21 @@ class JDLProcessor {
             return [baseName, { clientFramework, serverPort, applicationIndex, devServerPort }];
           })
         );
+
+        const gatewayClientFramework = applicationWithEntities.config.clientFramework;
```
Comment (mshima) [inline cli/import-jdl.js:444]: Move to line 434:
```
const { serverPort: gatewayServerPort, clientFramework: gatewayClientFramework } = applicationWithEntities.config;
```

File: cli/import-jdl.js [hunk 2]
```diff
@@ -440,6 +440,21 @@ class JDLProcessor {
...
+
+        const differentClientFrameworksUsed = !usedClientFrameworks.every(framework => gatewayClientFramework === framework);
+
+        if (differentClientFrameworksUsed) {
+          throw Error(
+            `Using different client frameworks in microfrontends is not supported. Tried to use: ${gatewayClientFramework} with ${usedClientFrameworks.join(
+              ', '
+            )}`
+          );
+        }
```
Comment (mshima) [inline cli/import-jdl.js:457]: Please reuse the other loop.
Comparing clientFramework from the microservice with the gateway is enough.
No need to extract every value for the log.

**Review by mshima — CHANGES_REQUESTED** [review]
Should be simplified.

File: cli/import-jdl.js (hunk 2 above)
Comment (mshima) [reply cli/import-jdl.js]: > eslint seems to not like it.

Should be forEach instead of map.

File: cli/import-jdl.js (hunk 1 above)
Comment (emilpaw) [reply cli/import-jdl.js]: Done

File: cli/import-jdl.js [hunk 3]
```diff
@@ -431,7 +431,7 @@ class JDLProcessor {
           applicationWithEntities.config.baseName !== baseName &&
           applicationWithEntities.entities.find(entity => entity.microserviceName === baseName)
```
Comment (mshima) [inline cli/import-jdl.js:432]: Add
```js
      const microfrontends = applicationWithEntities.config.microfrontends || [];
```

```suggestion
          (
            applicationWithEntities.entities.find(entity => entity.microserviceName === baseName) ||
            microfrontends.includes(baseName)
          )
```

**Review by mshima — CHANGES_REQUESTED** [review]
Unfortunately this won't work with main after https://github.com/jhipster/generator-jhipster/pull/18254 using new jdl microfrontend notation.
Old notation will be dropped at https://github.com/jhipster/generator-jhipster/pull/18632.
The relatedApplications needs to use microfrontends baseNames.

File: cli/import-jdl.js (hunk 3 above)
Comment (emilpaw) [reply cli/import-jdl.js]: Thanks. Done.

**Review by mshima — APPROVED** [review]

**Comment by emilpaw** [comment]
Bug bounty claimed: https://opencollective.com/generator-jhipster/expenses/115032 Thanks :)

**Comment by DanielFran** [comment]
@emilpaw approved

**Pull request merged**